File: src/sdks/ios/Sources/FraudTracker/FraudTracker.swift
import Foundation

/// Signal types captured by the FraudTracker
public enum SignalType: String, Codable {
    case touch = "touch"
    case gesture = "gesture"
    case accelerometer = "accelerometer"
    case gyroscope = "gyroscope"
    case device = "device"
    case appLifecycle = "app_lifecycle"
    case jailbreak = "jailbreak"
}

/// A captured behavioral signal
public struct Signal: Codable {
    public let type: SignalType
    public let timestamp: TimeInterval
    public let payload: [String: AnyCodable]
    
    public init(type: SignalType, timestamp: TimeInterval = Date().timeIntervalSince1970 * 1000, payload: [String: AnyCodable]) {
        self.type = type
        self.timestamp = timestamp
        self.payload = payload
    }
}

/// Session information
public struct Session {
    public let id: String
    public let clientId: String
    public let startedAt: TimeInterval
    public let deviceFingerprint: String
}

/// Configuration for FraudTracker
public struct FraudTrackerConfig {
    public let endpoint: String
    public let clientId: String
    public var batchSize: Int = 50
    public var flushInterval: TimeInterval = 0.5
    public var debug: Bool = false
    
    public init(endpoint: String, clientId: String) {
        self.endpoint = endpoint
        self.clientId = clientId
    }
}

/// Main FraudTracker class for iOS
public class FraudTracker {
    
    public static let shared = FraudTracker()
    
    private var config: FraudTrackerConfig?
    private var session: Session?
    private var signalBuffer: [Signal] = []
    private var flushTimer: Timer?
    private var isInitialized = false
    
    private init() {}
    
    // MARK: - Public API
    
    /// Initialize the tracker with configuration
    public func initialize(config: FraudTrackerConfig) {
        guard !isInitialized else {
            log("Already initialized")
            return
        }
        
        self.config = config
        self.session = createSession()
        self.isInitialized = true
        
        attachObservers()
        startFlushTimer()
        captureDeviceInfo()
        checkJailbreak()
        
        log("Initialized with session: \(session?.id ?? "unknown")")
    }
    
    /// Capture a custom signal
    public func capture(type: SignalType, payload: [String: AnyCodable]) {
        guard isInitialized else {
            print("[FraudTracker] Not initialized. Call initialize() first.")
            return
        }
        
        let signal = Signal(type: type, payload: payload)
        signalBuffer.append(signal)
        
        if let config = config, signalBuffer.count >= config.batchSize {
            flush()
        }
    }
    
    /// Complete the session and flush remaining signals
    public func complete(completion: (() -> Void)? = nil) {
        guard isInitialized, session != nil else { return }
        
        stopFlushTimer()
        detachObservers()
        flush()
        sendComplete()
        
        log("Session completed: \(session?.id ?? "unknown")")
        session = nil
        isInitialized = false
        
        completion?()
    }
    
    /// Get current session ID
    public func getSessionId() -> String? {
        return session?.id
    }
    
    // MARK: - Private Methods
    
    private func createSession() -> Session {
        return Session(
            id: UUID().uuidString,
            clientId: config?.clientId ?? "",
            startedAt: Date().timeIntervalSince1970 * 1000,
            deviceFingerprint: generateFingerprint()
        )
    }
    
    private func generateFingerprint() -> String {
        // Basic fingerprint - will be enhanced
        let device = UIDevice.current
        let components = [
            device.model,
            device.systemName,
            device.systemVersion,
            UIScreen.main.bounds.width.description,
            UIScreen.main.bounds.height.description,
            TimeZone.current.identifier
        ]
        return components.joined(separator: "|").hash.description
    }
    
    private func attachObservers() {
        // App lifecycle
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(appDidBecomeActive),
            name: UIApplication.didBecomeActiveNotification,
            object: nil
        )
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(appWillResignActive),
            name: UIApplication.willResignActiveNotification,
            object: nil
        )
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(appDidEnterBackground),
            name: UIApplication.didEnterBackgroundNotification,
            object: nil
        )
        
        // TODO: Add motion/accelerometer observers in Phase 1
    }
    
    private func detachObservers() {
        NotificationCenter.default.removeObserver(self)
    }
    
    @objc private func appDidBecomeActive() {
        capture(type: .appLifecycle, payload: ["state": AnyCodable("active")])
    }
    
    @objc private func appWillResignActive() {
        capture(type: .appLifecycle, payload: ["state": AnyCodable("inactive")])
    }
    
    @objc private func appDidEnterBackground() {
        capture(type: .appLifecycle, payload: ["state": AnyCodable("background")])
        flush()
    }
    
    private func captureDeviceInfo() {
        let device = UIDevice.current
        let screen = UIScreen.main
        
        capture(type: .device, payload: [
            "model": AnyCodable(device.model),
            "systemName": AnyCodable(device.systemName),
            "systemVersion": AnyCodable(device.systemVersion),
            "identifierForVendor": AnyCodable(device.identifierForVendor?.uuidString ?? "unknown"),
            "screenWidth": AnyCodable(screen.bounds.width),
            "screenHeight": AnyCodable(screen.bounds.height),
            "screenScale": AnyCodable(screen.scale),
            "timezone": AnyCodable(TimeZone.current.identifier),
            "locale": AnyCodable(Locale.current.identifier)
        ])
    }
    
    private func checkJailbreak() {
        var isJailbroken = false
        
        #if !targetEnvironment(simulator)
        // Check for common jailbreak paths
        let paths = [
            "/Applications/Cydia.app",
            "/Library/MobileSubstrate/MobileSubstrate.dylib",
            "/bin/bash",
            "/usr/sbin/sshd",
            "/etc/apt",
            "/private/var/lib/apt/"
        ]
        
        for path in paths {
            if FileManager.default.fileExists(atPath: path) {
                isJailbroken = true
                break
            }
        }
        
        // Check if app can write to private directory
        let testPath = "/private/jailbreak_test.txt"
        do {
            try "test".write(toFile: testPath, atomically: true, encoding: .utf8)
            try FileManager.default.removeItem(atPath: testPath)
            isJailbroken = true
        } catch {
            // Expected on non-jailbroken device
        }
        #endif
        
        capture(type: .jailbreak, payload: ["detected": AnyCodable(isJailbroken)])
    }
    
    // MARK: - Network
    
    private func startFlushTimer() {
        guard let config = config else { return }
        
        flushTimer = Timer.scheduledTimer(withTimeInterval: config.flushInterval, repeats: true) { [weak self] _ in
            if let count = self?.signalBuffer.count, count > 0 {
                self?.flush()
            }
        }
    }
    
    private func stopFlushTimer() {
        flushTimer?.invalidate()
        flushTimer = nil
    }
    
    private func flush() {
        guard let session = session, let config = config, !signalBuffer.isEmpty else { return }
        
        let signals = signalBuffer
        signalBuffer = []
        
        let url = URL(string: "\(config.endpoint)/api/v1/sessions/\(session.id)/signals")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let payload: [String: Any] = [
            "sessionId": session.id,
            "signals": signals.map { signal in
                [
                    "type": signal.type.rawValue,
                    "timestamp": signal.timestamp,
                    "payload": signal.payload
                ] as [String : Any]
            }
        ]
        
        request.httpBody = try? JSONSerialization.data(withJSONObject: payload)
        
        URLSession.shared.dataTask(with: request) { [weak self] _, response, error in
            if let error = error {
                self?.log("Flush failed: \(error)")
                // Re-add signals for retry
                self?.signalBuffer.insert(contentsOf: signals, at: 0)
            } else {
                self?.log("Flushed \(signals.count) signals")
            }
        }.resume()
    }
    
    private func sendComplete() {
        guard let session = session, let config = config else { return }
        
        let url = URL(string: "\(config.endpoint)/api/v1/sessions/\(session.id)/complete")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: ["completedAt": Date().timeIntervalSince1970 * 1000])
        
        URLSession.shared.dataTask(with: request).resume()
    }
    
    private func log(_ message: String) {
        if config?.debug == true {
            print("[FraudTracker] \(message)")
        }
    }
}

// MARK: - AnyCodable Helper

public struct AnyCodable: Codable {
    public let value: Any
    
    public init(_ value: Any) {
        self.value = value
    }
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        
        if let bool = try? container.decode(Bool.self) {
            value = bool
        } else if let int = try? container.decode(Int.self) {
            value = int
        } else if let double = try? container.decode(Double.self) {
            value = double
        } else if let string = try? container.decode(String.self) {
            value = string
        } else if let array = try? container.decode([AnyCodable].self) {
            value = array.map { $0.value }
        } else if let dict = try? container.decode([String: AnyCodable].self) {
            value = dict.mapValues { $0.value }
        } else {
            value = NSNull()
        }
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        
        switch value {
        case let bool as Bool:
            try container.encode(bool)
        case let int as Int:
            try container.encode(int)
        case let double as Double:
            try container.encode(double)
        case let string as String:
            try container.encode(string)
        case let array as [Any]:
            try container.encode(array.map { AnyCodable($0) })
        case let dict as [String: Any]:
            try container.encode(dict.mapValues { AnyCodable($0) })
        default:
            try container.encodeNil()
        }
    }
}

File: src/sdks/android/fraudtracker/src/main/kotlin/com/fraud/tracker/FraudTracker.kt
package com.fraud.tracker

import android.content.Context
import android.os.Build
import android.util.DisplayMetrics
import android.view.WindowManager
import kotlinx.coroutines.*
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedOutputStream
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Signal types captured by the FraudTracker
 */
enum class SignalType(val value: String) {
    TOUCH("touch"),
    GESTURE("gesture"),
    ACCELEROMETER("accelerometer"),
    GYROSCOPE("gyroscope"),
    DEVICE("device"),
    APP_LIFECYCLE("app_lifecycle"),
    ROOT_DETECTION("root_detection")
}

/**
 * A captured behavioral signal
 */
data class Signal(
    val type: SignalType,
    val timestamp: Long = System.currentTimeMillis(),
    val payload: Map<String, Any?>
)

/**
 * Session information
 */
data class Session(
    val id: String,
    val clientId: String,
    val startedAt: Long,
    val deviceFingerprint: String
)

/**
 * Configuration for FraudTracker
 */
data class FraudTrackerConfig(
    val endpoint: String,
    val clientId: String,
    val batchSize: Int = 50,
    val flushIntervalMs: Long = 500,
    val debug: Boolean = false
)

/**
 * Main FraudTracker class for Android
 */
class FraudTracker private constructor() {

    companion object {
        @Volatile
        private var instance: FraudTracker? = null

        fun getInstance(): FraudTracker {
            return instance ?: synchronized(this) {
                instance ?: FraudTracker().also { instance = it }
            }
        }
    }

    private var config: FraudTrackerConfig? = null
    private var session: Session? = null
    private var signalBuffer = CopyOnWriteArrayList<Signal>()
    private var isInitialized = false
    private var flushJob: Job? = null
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    // ─────────────────────────────────────────────────────────────
    // Public API
    // ─────────────────────────────────────────────────────────────

    /**
     * Initialize the tracker with configuration
     */
    fun initialize(context: Context, config: FraudTrackerConfig) {
        if (isInitialized) {
            log("Already initialized")
            return
        }

        this.config = config
        this.session = createSession(context)
        this.isInitialized = true

        startFlushTimer()
        captureDeviceInfo(context)
        checkRoot()

        log("Initialized with session: ${session?.id}")
    }

    /**
     * Capture a custom signal
     */
    fun capture(type: SignalType, payload: Map<String, Any?>) {
        if (!isInitialized) {
            println("[FraudTracker] Not initialized. Call initialize() first.")
            return
        }

        val signal = Signal(type = type, payload = payload)
        signalBuffer.add(signal)

        config?.let {
            if (signalBuffer.size >= it.batchSize) {
                flush()
            }
        }
    }

    /**
     * Complete the session and flush remaining signals
     */
    fun complete(callback: (() -> Unit)? = null) {
        if (!isInitialized || session == null) return

        stopFlushTimer()
        flush()
        sendComplete()

        log("Session completed: ${session?.id}")
        session = null
        isInitialized = false

        callback?.invoke()
    }

    /**
     * Get current session ID
     */
    fun getSessionId(): String? = session?.id

    /**
     * Handle app lifecycle - call from Activity
     */
    fun onResume() {
        capture(SignalType.APP_LIFECYCLE, mapOf("state" to "resumed"))
    }

    fun onPause() {
        capture(SignalType.APP_LIFECYCLE, mapOf("state" to "paused"))
    }

    fun onStop() {
        capture(SignalType.APP_LIFECYCLE, mapOf("state" to "stopped"))
        flush()
    }

    // ─────────────────────────────────────────────────────────────
    // Private Methods
    // ─────────────────────────────────────────────────────────────

    private fun createSession(context: Context): Session {
        return Session(
            id = UUID.randomUUID().toString(),
            clientId = config?.clientId ?: "",
            startedAt = System.currentTimeMillis(),
            deviceFingerprint = generateFingerprint(context)
        )
    }

    private fun generateFingerprint(context: Context): String {
        val components = listOf(
            Build.MODEL,
            Build.MANUFACTURER,
            Build.VERSION.SDK_INT.toString(),
            getScreenSize(context),
            TimeZone.getDefault().id
        )
        return components.joinToString("|").hashCode().toString()
    }

    private fun getScreenSize(context: Context): String {
        val windowManager = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val metrics = DisplayMetrics()
        @Suppress("DEPRECATION")
        windowManager.defaultDisplay.getMetrics(metrics)
        return "${metrics.widthPixels}x${metrics.heightPixels}"
    }

    private fun captureDeviceInfo(context: Context) {
        val windowManager = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val metrics = DisplayMetrics()
        @Suppress("DEPRECATION")
        windowManager.defaultDisplay.getMetrics(metrics)

        capture(SignalType.DEVICE, mapOf(
            "model" to Build.MODEL,
            "manufacturer" to Build.MANUFACTURER,
            "brand" to Build.BRAND,
            "device" to Build.DEVICE,
            "sdkVersion" to Build.VERSION.SDK_INT,
            "release" to Build.VERSION.RELEASE,
            "screenWidth" to metrics.widthPixels,
            "screenHeight" to metrics.heightPixels,
            "screenDensity" to metrics.density,
            "timezone" to TimeZone.getDefault().id,
            "locale" to Locale.getDefault().toString()
        ))
    }

    private fun checkRoot() {
        var isRooted = false

        // Check for common root paths
        val paths = listOf(
            "/system/app/Superuser.apk",
            "/sbin/su",
            "/system/bin/su",
            "/system/xbin/su",
            "/data/local/xbin/su",
            "/data/local/bin/su",
            "/system/sd/xbin/su",
            "/system/bin/failsafe/su",
            "/data/local/su",
            "/su/bin/su"
        )

        for (path in paths) {
            if (File(path).exists()) {
                isRooted = true
                break
            }
        }

        // Check build tags
        if (Build.TAGS?.contains("test-keys") == true) {
            isRooted = true
        }

        capture(SignalType.ROOT_DETECTION, mapOf("detected" to isRooted))
    }

    // ─────────────────────────────────────────────────────────────
    // Network
    // ─────────────────────────────────────────────────────────────

    private fun startFlushTimer() {
        flushJob = scope.launch {
            while (isActive) {
                delay(config?.flushIntervalMs ?: 500)
                if (signalBuffer.isNotEmpty()) {
                    flush()
                }
            }
        }
    }

    private fun stopFlushTimer() {
        flushJob?.cancel()
        flushJob = null
    }

    private fun flush() {
        val currentSession = session ?: return
        val currentConfig = config ?: return
        if (signalBuffer.isEmpty()) return

        val signals = signalBuffer.toList()
        signalBuffer.clear()

        scope.launch {
            try {
                val url = URL("${currentConfig.endpoint}/api/v1/sessions/${currentSession.id}/signals")
                val connection = url.openConnection() as HttpURLConnection
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/json")
                connection.doOutput = true

                val payload = JSONObject().apply {
                    put("sessionId", currentSession.id)
                    put("signals", JSONArray().apply {
                        signals.forEach { signal ->
                            put(JSONObject().apply {
                                put("type", signal.type.value)
                                put("timestamp", signal.timestamp)
                                put("payload", JSONObject(signal.payload))
                            })
                        }
                    })
                }

                BufferedOutputStream(connection.outputStream).use { out ->
                    out.write(payload.toString().toByteArray())
                    out.flush()
                }

                val responseCode = connection.responseCode
                if (responseCode in 200..299) {
                    log("Flushed ${signals.size} signals")
                } else {
                    log("Flush failed with code $responseCode")
                    signalBuffer.addAll(0, signals)
                }

                connection.disconnect()
            } catch (e: Exception) {
                log("Flush error: ${e.message}")
                signalBuffer.addAll(0, signals)
            }
        }
    }

    private fun sendComplete() {
        val currentSession = session ?: return
        val currentConfig = config ?: return

        scope.launch {
            try {
                val url = URL("${currentConfig.endpoint}/api/v1/sessions/${currentSession.id}/complete")
                val connection = url.openConnection() as HttpURLConnection
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/json")
                connection.doOutput = true

                val payload = JSONObject().apply {
                    put("completedAt", System.currentTimeMillis())
                }

                BufferedOutputStream(connection.outputStream).use { out ->
                    out.write(payload.toString().toByteArray())
                    out.flush()
                }

                connection.disconnect()
            } catch (e: Exception) {
                log("Complete error: ${e.message}")
            }
        }
    }

    private fun log(message: String) {
        if (config?.debug == true) {
            println("[FraudTracker] $message")
        }
    }
}
